README.md
# aligned-pitch-filter

Python tools to filter/correct a pitch track according to the alignment with the relevant score.

Currently the algorithm is able to correct the octave errors by taking pitch heights of the aligned notes as the reference.

Usage
=======
All the relevant data can be easily obtained:

```python
from alignedpitchfilter.AlignedPitchFilter import AlignedPitchFilter

alignedPitchFilter = AlignedPitchFilter()
pitch_corrected, notes_corrected, synth_pitch = alignedPitchFilter.filter(pitch, notes)
```

The inputs are:
```python
# pitch 		  :	an n-by-2 matrix, where the values in the first column are 
#					the timestamps and the values in the second column are frequency 
#					values
# notes			  :	list of note structure. This is read from the alignedNotes.json 
#					output from https://github.com/sertansenturk/fragmentLinker repository 
```

The outputs are:
```python
# pitch_corrected :	The octave corrected pitch track. The size is equal to the size of pitch
# notes_corrected : Aligned notes which are removed according to the pitch filtering (e.g. notes with zero duration)
# synth_pitch	  :	Synthetic pitch track from the notes input that is used for octave correction
```

You can refer to [filter_pitch.ipynb](https://github.com/sertansenturk/alignedpitchfilter/blob/master/filter_pitch.ipynb) for an interactive demo. Moreover, [extract_pitch.ipynb](https://github.com/sertansenturk/alignedpitchfilter/blob/master/extract_pitch.ipynb) shows how to extract the predominant melody using [tomato](https://github.com/sertansenturk/tomato/tree/master/tomato).

Installation
============

If you want to install the repository, it is recommended to install the package and dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then
install the repository like this instead

    pip install -e .

Now you can install the rest of the dependencies:

    pip install -r requirements

Authors
-------
Sertan Senturk
contact@sertansenturk.com

Reference
-------
Thesis


Changelog.md
#### alignedpitchfilter v1.1
 - Simplified octave correction
 - Added plot function
 - Adjustments to class usage
 - Minor refactoring and clean ups

#### alignedpitchfilter v1.0
 - First public release


alignedpitchfilter/AlignedPitchFilter.py
import numpy as np
import copy
import matplotlib.pyplot as plt


class AlignedPitchFilter(object):
    def __init__(self):
        self.max_boundary_tol = 3  # seconds

    def filter(self, pitch, notes):
        # IMPORTANT: In the audio-score alignment step, the pitch value
        # of each note is computed from the theoretical pitch distance
        # from the tonic and de-normalized according to the tonic
        # frequency of the performance. The value is not computed from
        # THE PITCH TRAJECTORY OF THE NOTE; it is just the THEORETICAL
        # FREQUENCY OF THE NOTE SYMBOL ACCORDING TO THE TONIC FREQUENY
        # The performed stable pitch of the note will be computed in the
        # aligned-note-models

        pitch_corrected = np.copy(pitch)
        notes_corrected = copy.deepcopy(notes)

        # remove skipped notes
        notes_corrected = ([n for n in notes_corrected
                            if not n['Interval'][0] == n['Interval'][1]])

        # remove rests
        notes_corrected = [n for n in notes_corrected if
                           n['TheoreticalPitch']['Value']]

        # group the notes into sections
        synth_pitch = self._notes_to_synth_pitch(
            notes_corrected, pitch_corrected[:, 0])

        # octave correction
        for i, sp in enumerate(synth_pitch[:, 1]):
            pitch_corrected[i][1] = self._move_to_closest_octave(
                pitch_corrected[i][1], sp)

        for nc in notes_corrected:
            trajectory = np.vstack(
                p[1] for p in pitch_corrected
                if nc['Interval'][0] <= p[0] <= nc['Interval'][1])
            nc['PerformedPitch']['Value'] = np.median(trajectory).tolist()

        return pitch_corrected, notes_corrected, synth_pitch.tolist()

    def _notes_to_synth_pitch(self, notes, time_stamps):
        synth_pitch = np.array([0] * len(time_stamps))

        for i in range(0, len(notes)):
            prevlabel = ([] if i == 0 else
                         notes[i - 1]['Label'].split('--')[0])
            label = notes[i]['Label'].split('--')[0]
            nextlabel = ([] if i == len(notes) - 1 else
                         notes[i + 1]['Label'].split('--')[0])

            # pre interpolation start time on the first note
            if not prevlabel:
                startidx = self._find_closest_sample_idx(
                    notes[i]['Interval'][0] - self.max_boundary_tol,
                    time_stamps)
            elif not label == prevlabel:
                # post interpolation start time on a group start

                # recalculate the end time of the previous
                tempstartidx = self._find_closest_sample_idx(
                    notes[i]['Interval'][0], time_stamps)
                prevendidx = self._find_closest_sample_idx(
                    notes[i - 1]['Interval'][1] + self.max_boundary_tol,
                    time_stamps[:tempstartidx])

                startidx = prevendidx + self._find_closest_sample_idx(
                    notes[i]['Interval'][0] - self.max_boundary_tol,
                    time_stamps[prevendidx:]) + 1
            else:  # no pre interpolation
                startidx = self._find_closest_sample_idx(
                    notes[i]['Interval'][0], time_stamps)

            if not nextlabel:
                # post interpolation end time on the last note
                endidx = self._find_closest_sample_idx(
                    notes[i]['Interval'][1] + self.max_boundary_tol,
                    time_stamps)
            elif not label == nextlabel:
                # post interpolation end time on a group end
                nextstartidx = self._find_closest_sample_idx(
                    notes[i + 1]['Interval'][0], time_stamps)
                endidx = self._find_closest_sample_idx(
                    notes[i]['Interval'][1] + self.max_boundary_tol,
                    time_stamps[:nextstartidx])
            else:
                # post interpolation within a group
                nextstartidx = self._find_closest_sample_idx(
                    notes[i + 1]['Interval'][0], time_stamps)
                endidx = nextstartidx - 1

            synth_pitch[startidx:endidx + 1] = \
                notes[i]['TheoreticalPitch']['Value']

        # add time_stamps
        synth_pitch = np.transpose(np.vstack((time_stamps, synth_pitch)))

        return synth_pitch

    @staticmethod
    def _find_closest_sample_idx(val, sample_vals):
        return np.argmin(abs(sample_vals - val))

    @classmethod
    def _move_to_closest_octave(cls, pp, sp):
        minpp = pp
        if not (pp in [0, np.nan] or sp in [0, np.nan]):
            cent_diff = cls._hz2cent(pp, sp)
            octave_cands = [cent_diff, cent_diff - 1200]
            cand_dist = [abs(oc) for oc in octave_cands]
            closest_cent_diff = octave_cands[cand_dist.index(min(cand_dist))]

            minpp = cls._cent2hz(closest_cent_diff, sp)

        return minpp

    @staticmethod
    def _cent2hz(cent_val, ref_hz):
        try:
            return ref_hz * 2 ** (cent_val / 1200.0)
        except TypeError:  # _NaN_; rest
            return None

    @staticmethod
    def _hz2cent(val, ref_hz):
        return 1200.0 * np.log2(val / ref_hz)

    @staticmethod
    def _decompose_into_chunks(pitch, bottom_limit=0.7, upper_limit=1.3):
        pitch_chunks = []
        temp_pitch = np.array([])
        # starts at the first sample
        for i in range(1, len(pitch) - 1):
            # ignore chunks with 0 frequency values
            if pitch[i][1] == 0:
                pass
            # non-zero chunks
            else:
                interval = float(pitch[i + 1][1]) / float(pitch[i][1])
                temp_pitch = (np.vstack((temp_pitch, pitch[i]))
                              if temp_pitch.size > 0 else pitch[i])

                if not bottom_limit < interval < upper_limit:
                    if len(temp_pitch) > 0:
                        pitch_chunks.append(temp_pitch)
                    temp_pitch = np.array([])
        if len(temp_pitch) > 0:
            pitch_chunks.append(temp_pitch)

        return pitch_chunks

    @staticmethod
    def plot(pitch, pitch_corrected, notes_corrected):
        # remove zeros for plotting
        pitch_plot = np.copy(pitch)
        pitch_plot[pitch_plot[:, 1] == 0, 1] = np.NAN

        pitch_corrected_plot = np.copy(pitch_corrected)
        pitch_corrected_plot[pitch_corrected_plot[:, 1] == 0, 1] = np.NAN

        fig, ax = plt.subplots()

        # plot pitch tracks
        ax.plot(pitch_plot[:, 0], pitch_plot[:, 1], 'g', label='Pitch',
                alpha=0.7)
        ax.plot(pitch_corrected_plot[:, 0], pitch_corrected_plot[:, 1], 'b',
                label=u'Corrected Pitch')

        plt.xlabel('Time (sec)')
        plt.ylabel('Frequency (Hz)')

        plt.grid(True)

        # plot notes except the last one
        for note in notes_corrected[:-1]:
            ax.plot(note['Interval'], [note['PerformedPitch']['Value'],
                                       note['PerformedPitch']['Value']],
                    'r', alpha=0.4, linewidth=4)

        # plot last note for labeling
        dmy_note = notes_corrected[-1]
        ax.plot(dmy_note['Interval'], [dmy_note['PerformedPitch']['Value'],
                                       dmy_note['PerformedPitch']['Value']],
                'r', label=u'Aligned Notes', alpha=0.4, linewidth=4)

        # set y axis limits
        pitch_vals = np.hstack((pitch_plot[:, 1], pitch_corrected_plot[:, 1]))
        pitch_vals = pitch_vals[~np.isnan(pitch_vals)]

        min_y = np.min(pitch_vals)
        max_y = np.max(pitch_vals)
        range_y = max_y - min_y

        ax.set_ylim([min_y - range_y * 0.1, max_y + range_y * 0.1])

        # set x axis limits
        time_vals = np.hstack((pitch_plot[:, 0], pitch_corrected_plot[:, 0]))

        min_x = np.min(time_vals)
        max_x = np.max(time_vals)

        ax.set_xlim([min_x, max_x])

        # place legend
        ax.legend(loc='upper right')


alignedpitchfilter/alignedpitchfilter.py
import numpy as np
import copy


def corect_octave_errors(pitch, notes):
    # IMPORTANT: In the audio-score alignment step, the pitch value
    # of each note is computed from the theoretical pitch distance
    # from the tonic and de-normalized according to the tonic
    # frequency of the performance. The value is not computed from
    # THE PITCH TRAJECTORY OF THE NOTE; it is just the THEORETICAL
    # FREQUENCY OF THE NOTE SYMBOL ACCORDING TO THE TONIC FREQUENY
    # The performed stable pitch of the note will be computed in the
    # aligned-note-models

    notes_corrected = copy.deepcopy(notes)

    # remove skipped notes
    notes_corrected = ([n for n in notes_corrected
                        if not n['Interval'][0] == n['Interval'][1]])

    # remove rests
    notes_corrected = [n for n in notes_corrected if
                       n['TheoreticalPitch']['Value']]

    # group the notes into sections
    synth_pitch = notes_to_synth_pitch(notes_corrected, pitch[:, 0])

    # octave correction
    pitch_corrected = np.copy(pitch)
    for i, sp in enumerate(synth_pitch):
        pitch_corrected[i][1] = move_to_same_octave(pitch[i][1], sp)

    for nc in notes_corrected:
        trajectory = np.vstack(p[1] for p in pitch_corrected
                               if
                               nc['Interval'][0] <= p[0] <= nc['Interval'][1])
        nc['PerformedPitch']['Value'] = np.median(trajectory).tolist()
    # import pdb
    # pdb.set_trace()
    return pitch_corrected, synth_pitch, notes_corrected


def notes_to_synth_pitch(notes, time_stamps, max_boundary_tol=6):
    synth_pitch = np.array([0] * len(time_stamps))

    for i in range(0, len(notes)):
        prevlabel = ([] if i == 0 else
                     notes[i - 1]['Label'].split('--')[0])
        label = notes[i]['Label'].split('--')[0]
        nextlabel = ([] if i == len(notes) - 1 else
                     notes[i + 1]['Label'].split('--')[0])

        # pre interpolation start time on the first note
        if not prevlabel:
            startidx = find_closest_sample_idx(
                notes[i]['Interval'][0] - max_boundary_tol, time_stamps)
        elif not label == prevlabel:
            # post interpolation start time on a group start

            # recalculate the end time of the previous
            tempstartidx = find_closest_sample_idx(
                notes[i]['Interval'][0], time_stamps)
            prevendidx = find_closest_sample_idx(
                notes[i - 1]['Interval'][1] + max_boundary_tol,
                time_stamps[:tempstartidx])

            startidx = prevendidx + find_closest_sample_idx(
                notes[i]['Interval'][0] - max_boundary_tol,
                time_stamps[prevendidx:]) + 1
        else:  # no pre interpolation
            startidx = find_closest_sample_idx(
                notes[i]['Interval'][0], time_stamps)

        if not nextlabel:
            # post interpolation end time on the last note
            endidx = find_closest_sample_idx(
                notes[i]['Interval'][1] + max_boundary_tol, time_stamps)
        elif not label == nextlabel:
            # post interpolation end time on a group end
            nextstartidx = find_closest_sample_idx(
                notes[i + 1]['Interval'][0], time_stamps)
            endidx = find_closest_sample_idx(
                notes[i]['Interval'][1] + max_boundary_tol,
                time_stamps[:nextstartidx])
        else:
            # post interpolation within a group
            nextstartidx = find_closest_sample_idx(
                notes[i + 1]['Interval'][0], time_stamps)
            endidx = nextstartidx - 1

        synth_pitch[startidx:endidx + 1] = \
            notes[i]['TheoreticalPitch']['Value']

    return synth_pitch


def find_closest_sample_idx(val, sample_vals):
    return np.argmin(abs(sample_vals - val))


def move_to_same_octave(pp, sp):
    minpp = pp
    if not (pp == 0 or sp == 0):
        direction = 1 if sp > pp else -1

        prev_cent_diff = 1000000  # assign an absurd number
        decr = True
        while decr:
            cent_diff = abs(hz2cent(pp, sp))
            if prev_cent_diff > cent_diff:
                minpp = pp
                pp *= 2 ** direction
                prev_cent_diff = cent_diff
            else:
                decr = False

    return minpp


def cent2hz(cent_val, ref_hz):
    try:
        return ref_hz * 2 ** (cent_val / 1200.0)
    except TypeError:  # _NaN_; rest
        return None


def hz2cent(val, ref_hz):
    return 1200.0 * np.log2(val / ref_hz)


def decompose_into_chunks(pitch, bottom_limit=0.7, upper_limit=1.3):
    pitch_chunks = []
    temp_pitch = np.array([])
    # starts at the first sample
    for i in range(1, len(pitch) - 1):
        # ignore chunks with 0 frequency values
        if pitch[i][1] == 0:
            pass
        # non-zero chunks
        else:
            interval = float(pitch[i + 1][1]) / float(pitch[i][1])
            temp_pitch = (np.vstack((temp_pitch, pitch[i]))
                          if temp_pitch.size > 0 else pitch[i])

            if not bottom_limit < interval < upper_limit:
                if len(temp_pitch) > 0:
                    pitch_chunks.append(temp_pitch)
                temp_pitch = np.array([])
    if len(temp_pitch) > 0:
        pitch_chunks.append(temp_pitch)

    return pitch_chunks


filteralignedpitch/filteralignedpitch.py
import numpy as np
import pdb

def correctOctaveErrors(pitch, notes, tonic, ):
	# convert the symbolic pitch heights recorded in notes to Hz wrt tonic
	for note in notes:
		note['SymbolicPitch'] = note['Pitch']
		note['Pitch'] = {'Value': cent2hz(note['Pitch']['Value'],
			tonic['Value']),'Unit':'Hz'}

	# remove skipped notes
	notes = ([n for n in notes 
		if not n['Interval'][0] == n['Interval'][1]])

	# remove rests
	notes = [n for n in notes if n['Pitch']['Value']]

	# group the notes into sections
	synth_pitch = notes2synthPitch(notes, pitch[:,0])

	# octave correction
	pitch_corrected = np.copy(pitch)
	for i, sp in enumerate(synth_pitch):
		pitch_corrected[i][1] = move2sameOctave(pitch[i][1], sp)

	return pitch_corrected, synth_pitch, notes

def notes2synthPitch(notes, time_stamps, max_boundary_tol = 6):
	synthPitch = np.array([0] * len(time_stamps))

	for i in range(0,len(notes)):
		prevlabel = ([] if i == 0 else
			notes[i-1]['Label'].split('--')[0])
		label = notes[i]['Label'].split('--')[0]
		nextlabel = ([] if i == len(notes)-1 else
			notes[i+1]['Label'].split('--')[0])

		# pre interpolation start time on the first note
		if not prevlabel:
			startidx = find_closest_sample_idx(
				notes[i]['Interval'][0]-max_boundary_tol, time_stamps)
		elif not label == prevlabel:
			# post interpolation start time on a group start

			# recalculate the end time of the previous
			tempstartidx = find_closest_sample_idx(
				notes[i]['Interval'][0], time_stamps)
			prevendidx = find_closest_sample_idx(
				notes[i-1]['Interval'][1]+max_boundary_tol, 
				time_stamps[:tempstartidx])

			startidx = prevendidx+find_closest_sample_idx(
				notes[i]['Interval'][0]-max_boundary_tol, 
				time_stamps[prevendidx:])+1
		else:  # no pre interpolation
			startidx = find_closest_sample_idx(
				notes[i]['Interval'][0], time_stamps)

		if not nextlabel:
			# post interpolation end time on the last note
			endidx = find_closest_sample_idx(
				notes[i]['Interval'][1]+max_boundary_tol, time_stamps)
		elif not label == nextlabel:
			# post interpolation end time on a group end
			nextstartidx = find_closest_sample_idx(
				notes[i+1]['Interval'][0], time_stamps)
			endidx = find_closest_sample_idx(
				notes[i]['Interval'][1]+max_boundary_tol, 
				time_stamps[:nextstartidx])
		else:
			# post interpolation within a group
			nextstartidx = find_closest_sample_idx(
				notes[i+1]['Interval'][0], time_stamps)
			endidx = nextstartidx-1
		
		synthPitch[startidx:endidx+1] = notes[i]['Pitch']['Value']
	
	return synthPitch

def find_closest_sample_idx(val, sampleVals):
	return np.argmin(abs(sampleVals - val))

def move2sameOctave(pp, sp):
	minpp = pp
	if not(pp == 0 or sp == 0):
		direction = 1 if sp > pp else -1

		prev_cent_diff = 1000000 # assign an absurd number
		decr = True
		while decr:
			cent_diff = abs(hz2cent(pp,sp))
			if prev_cent_diff > cent_diff:
				minpp = pp
				pp = pp * (2 ** direction)
				prev_cent_diff = cent_diff
			else:
				decr = False

	return minpp

def cent2hz(centVal, refHz):
	try:
		return refHz * 2**(centVal/1200)
	except TypeError:  # _NaN_; rest
		return None

def hz2cent(val, refHz):
	return 1200 * np.log2(val/refHz)

def decompose_into_chunks(pitch, bottom_limit=0.7, upper_limit=1.3):
    """
    decomposes the given pitch track into the chunks.
    """

    pitch_chunks = []
    temp_pitch = np.array([])
    # starts at the first sample
    for i in range(1, len(pitch) - 1):
        # ignore chunks with 0 frequency values
        if pitch[i][1] == 0:
            pass
        # non-zero chunks
        else:
            interval = float(pitch[i + 1][1]) / float(pitch[i][1])
            temp_pitch = (np.vstack((temp_pitch, pitch[i])) 
            	if temp_pitch.size > 0 else pitch[i])

            if not bottom_limit < interval < upper_limit:
                if len(temp_pitch) > 0: pitch_chunks.append(temp_pitch)
                temp_pitch = np.array([])
    if len(temp_pitch) > 0: pitch_chunks.append(temp_pitch)

    return pitch_chunks